### PR TITLE
Notifications

### PR DESCRIPTION
**Pull request by lisset-orozco**
- [x]  Create notification table with a polymorphic field

- [x] Create the notification module and include it in the Post and Friendship model

- [x] Create job to send notifications

- [x] Merge with master

- [x]  Configuring sidekiq

- [x] Channel and job to send notifications

- [x] Show notifications in real time

- [x] Show notifications with polymorphic items

- [x] Show notifications with javascript

- [x] Mark notifications as viewed

### CODE REVIEW

File: app/controllers/notifications_controller.rb [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    respond_to do |format|
+      format.html
+      format.js
+    end
+  end
+
+  def update
+    @notification = Notification.find_id(params[:id])
+    message =
+      if @notification.update(notification_params)
```
Comment (joalbertg) [inline app/controllers/notifications_controller.rb]: refactor to a single line and use a ternary operator

File: app/decorators/notification_decorator.rb [hunk 2]
```diff
@@ -0,0 +1,13 @@
+class NotificationDecorator < Draper::Decorator
```
Comment (joalbertg) [inline app/decorators/notification_decorator.rb]: delete empty files

File: app/assets/stylesheets/notifications.scss [hunk 3]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the Notifications controller here.
```
Comment (joalbertg) [inline app/assets/stylesheets/notifications.scss]: delete empty files

File: app/assets/javascripts/notifications.coffee [hunk 4]
```diff
@@ -0,0 +1,3 @@
+# Place all the behaviors and hooks related to the matching controller here.
```
Comment (joalbertg) [inline app/assets/javascripts/notifications.coffee]: delete empty files

File: app/jobs/notification_sender_job.rb [hunk 5]
```diff
@@ -0,0 +1,13 @@
+# frozen_string_literal: true
+
+# notification job
+class NotificationSenderJob < ApplicationJob
+  queue_as :notifications
+
+  def perform(item)
+    # Do something later
+    item.user_ids.each do |user_id|
```
Comment (joalbertg) [inline app/jobs/notification_sender_job.rb]: refactor to only one line

File: app/models/friendship.rb [hunk 6]
```diff
@@ -37,6 +38,8 @@ class Friendship < ApplicationRecord
     end
   end
 
+  def user_ids; end
```
Comment (joalbertg) [inline app/models/friendship.rb]: ?

File: app/models/notification.rb [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+#  item_type  :string
+#  item_id    :integer
+#  viewed     :boolean          default(FALSE)
+#  created_at :datetime         not null
+#  updated_at :datetime         not null
+#
+# notification
+class Notification < ApplicationRecord
+  belongs_to :user
+  belongs_to :item, polymorphic: true
```
Comment (joalbertg) [inline app/models/notification.rb:18]: add line break

File: app/models/notification.rb [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+  after_commit { NotificationBroadcastJob.perform_later(self) }
```
Comment (joalbertg) [inline app/models/notification.rb]: add line break

Comment (joalbertg) [reply app/models/notification.rb]: avoid callbacks, possibly the solution server pattern.

**Review by joalbertg — CHANGES_REQUESTED** [review]

File: app/controllers/notifications_controller.rb (hunk 1 above)
Comment (lisset-orozco) [reply app/controllers/notifications_controller.rb]: Done!

File: app/assets/stylesheets/notifications.scss (hunk 3 above)
Comment (lisset-orozco) [reply app/assets/stylesheets/notifications.scss]: done!

File: app/assets/javascripts/notifications.coffee (hunk 4 above)
Comment (lisset-orozco) [reply app/assets/javascripts/notifications.coffee]: done!

File: app/jobs/notification_sender_job.rb (hunk 5 above)
Comment (lisset-orozco) [reply app/jobs/notification_sender_job.rb]: done

File: app/models/friendship.rb (hunk 6 above)
Comment (lisset-orozco) [reply app/models/friendship.rb]: done

File: app/models/notification.rb (hunk 7 above)
Comment (lisset-orozco) [reply app/models/notification.rb:18]: done

File: app/controllers/notifications_controller.rb (hunk 1 above)
Comment (lisset-orozco) [reply app/controllers/notifications_controller.rb]: Done!

File: app/decorators/notification_decorator.rb (hunk 2 above)
Comment (lisset-orozco) [reply app/decorators/notification_decorator.rb]: Done

File: app/models/notification.rb (hunk 8 above)
Comment (lisset-orozco) [reply app/models/notification.rb]: done

File: app/controllers/posts_controller.rb [hunk 9]
```diff
@@ -25,17 +25,16 @@ def edit; end
   # POST /posts
   # POST /posts.json
   def create
-    @post = current_user.posts.new(post_params)
+    @post = PostService.new(post_params, current_user)
```
Comment (joalbertg) [inline app/controllers/posts_controller.rb]: change name to `@post_service`

File: app/jobs/notification_sender_job.rb [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+  end
+
+  private
+
+  def generate(user_id, item)
+    NotificationService.new(item, user_id).create_object
+  end
+
+  def item(id, class_name)
+    return Post.find_by_id(id) if class_name == 'Post'
```
Comment (joalbertg) [inline app/jobs/notification_sender_job.rb]: change to meta programming!!!

File: app/services/notification_service.rb [hunk 11]
```diff
@@ -0,0 +1,30 @@
+# frozen_string_literal: true
+
+# service
+class NotificationService
+  def initialize(item, user_id)
+    @notification = Notification.new(item: item, user_id: user_id)
+  end
+
+  def create_object
+    if @notification.save
```
Comment (joalbertg) [inline app/services/notification_service.rb]: refactor with `unless`

File: app/services/post_service.rb [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+
+# service
+class PostService
+  def initialize(params, current_user)
+    @user = current_user
+    @post = @user.posts.new(params)
+  end
+
+  def create_object
+    if @post.save
```
Comment (joalbertg) [inline app/services/post_service.rb]: same here

**Review by joalbertg — CHANGES_REQUESTED** [review]

File: app/controllers/posts_controller.rb (hunk 9 above)
Comment (lisset-orozco) [reply app/controllers/posts_controller.rb]: done

File: app/jobs/notification_sender_job.rb (hunk 10 above)
Comment (lisset-orozco) [reply app/jobs/notification_sender_job.rb]: done

File: app/services/notification_service.rb (hunk 11 above)
Comment (lisset-orozco) [reply app/services/notification_service.rb]: done

File: app/services/post_service.rb (hunk 12 above)
Comment (lisset-orozco) [reply app/services/post_service.rb]: done

File: app/jobs/notification_sender_job.rb [hunk 13]
```diff
@@ -0,0 +1,32 @@
+# frozen_string_literal: true
+
+# notification job
+class NotificationSenderJob < ApplicationJob
+  queue_as :notifications
+
+  def perform(id, class_name)
```
Comment (joalbertg) [inline app/jobs/notification_sender_job.rb:7]: call only one method within `perform`, and move the private methods to another model class or reference service.

**Review by joalbertg — CHANGES_REQUESTED** [review]

File: app/jobs/notification_sender_job.rb (hunk 13 above)
Comment (lisset-orozco) [reply app/jobs/notification_sender_job.rb:7]: done!

**Review by joalbertg — APPROVED** [review]

**Pull request merged**